### ISSUE TITLE
shared propagation of "mount.entry" does not work

### ISSUE THREAD

**Issue by flx42** [lxc/lxc#2477]
Could probably be bundled in #2475

Let's assume I want to bind-mount `/run` as `shared` inside my container:
```
$ findmnt /run -o TARGET,OPT-FIELDS,PROPAGATION
TARGET OPT-FIELDS PROPAGATION
/run   shared:5   shared
```
Currently, we get the following:
```
$ sudo lxc-execute --define 'lxc.mount.entry = /run mnt/run none bind,shared,create=dir 0 0' -n bionic -- findmnt -o TARGET,OPT-FIELDS,PROPAGATION /mnt/run
TARGET   OPT-FIELDS          PROPAGATION
/mnt/run shared:237 master:5 shared,slave
```
This mount inside the container is a slave of the `/run` group (group 5). It should have been `shared:5`.
Doing a mount inside the container will not make the mount visible on the host.

I think what happens is:
- `remount_all_slave` will remount `/run` as `slave`.
- when bind-mounting `/run` inside the container's filesystem, the resulting mount point will be `slave` per the bind semantics of the [kernel documentation](https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt)
- `make_shared` on a slave mount will make it `shared and slave` per the state diagram in the documentation.

**Comment by brauner** [comment]
Yeah, we need to discuss this. There are multiple problems with all of this. One of the main problems is that `pivot_root()` will `no no` you when your rootfs is `MS_SHARED`. I've been thinking about this and I haven't come up with a clean and clear solution yet.

**Comment by brauner** [comment]
So suggestion welcome!

**Comment by brauner** [comment]
Hm, can you test whether this would work for LXC too?

**Comment by flx42** [comment]
Thinking about it more, it looks like this approach is better (slightly more fine-grained) but still has a major issue. Let's take the following example
```
lxc.mount.entry = /run/nvidia mnt/nvidia none bind,shared,create=dir 0 0
lxc.mount.entry = /home/fabecassis/test mnt/test none bind,shared,create=dir 0 0
```

If the parent mount is set to `private` (or `slave`). The behavior of these two lines above will depend on where the `rootfs` is:
* `/run/rootfs`: `/run` will be made `private`, the first mount can't be `shared`.
* `/home/fabecassis/rootfs`: `/home/ will be made `private`, the second mount can't be `shared`.
* `/tmp/rootfs`: `/tmp` or `/` will be made private, both mounts will be fine.

More generally, I'm not sure how we can reconcile these two constraints:
1. No mount must propagate to the host.
2. We can bind-mount any location with `shared` propagation.

For 1., we need to remove the default `shared` propagation on `/` or at least the parent mount, but for this mount there is no way back, you can't do 2.

**Comment by flx42** [comment]
Any idea @brauner?

**Comment by flx42** [comment]
My understanding of what runc is doing today:
```sh
ROOTFS="/home/fabecassis/rootfs"

mount --make-private /home # Parent mount of the rootfs

mount --bind "${ROOTFS}" "${ROOTFS}"
mount --make-private "${ROOTFS}"

mount --bind "/home/fabecassis/test" "$ROOTFS/mnt/test"

cd "${ROOTFS}"
mount -t proc none proc

pivot_root . mnt
mount --make-rslave mnt
umount -l mnt
exec chroot . findmnt -R -o TARGET,OPT-FIELDS,PROPAGATION /mnt/test
# Mount is in a new peer group, not the host one.
```

To run:
```
sudo unshare --mount --pid --fork --propagation=unchanged ./script.sh
```

@3XX0 suggested that we let the mount propagate on the host, and we clean it up, in the code above it's just a matter of doing the `mount --make-private /home` later:
```
ROOTFS="/home/fabecassis/rootfs"

mount --bind "${ROOTFS}" "${ROOTFS}"
mount --make-private "${ROOTFS}"

mount --bind "/home/fabecassis/test" "$ROOTFS/test"

mount --make-private /home # <- Now after the bind-mount.

cd "${ROOTFS}"
mount -t proc none proc

pivot_root . mnt
mount --make-rslave mnt
umount -l mnt

exec chroot . findmnt -o TARGET,OPT-FIELDS,PROPAGATION
```
To launch it:
```
unshare --mount --pid --fork --propagation=unchanged ./script4.sh
umount /home/fabecassis/rootfs
```
This makes things complicated if things goes wrong though.

**Comment by 3XX0** [comment]
I was more thinking of something like this:

```bash
sudo unshare -m --propagation=unchanged

WORKDIR=$(mktemp -d --tmpdir workdir.XXXXX)
mount --bind $WORKDIR $WORKDIR
mount --make-private $WORKDIR

# At this point we may have leaked $WORKDIR to the host but it doesn't really matter since it has a unique name, it's empty and it's temporary anyway.
# (alternatively we could use a well-known working directory like lxc.rootfs.mount but we would have to flock it or similar to prevent propagation races between multiple containers).

mkdir $WORKDIR/rootfs
mount --bind $ROOTFS $WORKDIR/rootfs
mount --make-slave $WORKDIR/rootfs
cd $WORKDIR/rootfs

# Now all the condition are fulfilled for pivot_root and nothing should have leaked to the host.

mount -t proc proc proc
pivot_root . mnt

# Unmount the working directory first (in case of shared propagation) and remove it, this should cleanup the host.

umount $WORKDIR
rm -rf /mnt/$WORKDIR

# Finally, unmount the oldroot as usual
mount --make-rslave mnt
umount -l mnt
```

**Comment by brauner** [comment]
Is this something either one of you can make a POC patch for?

**Comment by 3XX0** [comment]
https://github.com/3XX0/lxc/commit/ef541edfbcda6ca4ae30a9daee8d66cda44e89e8

Obviously hacky but you get the idea.

I didn't handle the `lxc_chroot` code path and I'm not sure what the `if (!rootfs->path) ...  MS_REC | MS_SLAVE` is about in `lxc_mount_rootfs`.

This also assumes that  `lxc.rootfs.mount` has one parent. In the case of the default `/usr/.../lxc/rootfs`, it's `lxc`

**Comment by brauner** [comment]
On Fri, Jul 20, 2018 at 07:39:33PM -0700, Jonathan Calmels wrote:
> https://github.com/3XX0/lxc/commit/ef541edfbcda6ca4ae30a9daee8d66cda44e89e8
> 
> Obviously hacky but you get the idea.
> 
> I didn't handle the `lxc_chroot` code path and I'm not sure what the `if (!rootfs->path) ...  MS_REC | MS_SLAVE` is about in `lxc_mount_rootfs`.

Well, when you share the rootfs with the host you really don't want to
propagate your mount and umounts. :)

> 
> This also assumes that  `lxc.rootfs.mount` has one parent. In the case of the default `/usr/.../lxc/rootfs`, it's `lxc`

Sigh, this is all going to be rather ugly. Let's see.

**Comment by 3XX0** [comment]
@brauner what do you suggest we do? :)

> Well, when you share the rootfs with the host you really don't want to
propagate your mount and umounts. :)

Shouldn't this be already handled by `remount_all_slave`?
And one might still want propagation to work no?

BTW, I commented out this function but that means that every bind mount entry should probably default to slave propagation otherwise the bind mount will inherit whatever is set on the host.

> Sigh, this is all going to be rather ugly. Let's see.

Yeah, the solution would be to create another directory under `lxc.rootfs.mount` but that's not ideal.

**Comment by brauner** [comment]
I've pinged Ram Pai who wrote the original `MS_SHARED` logic and blocked `pivot_root()` for that case. I want to hear why he did it since the commit message in the kernel doesn't explain it at all.

**Comment by brauner** [comment]
Once we have an explanation why this definitely can't be allowed we move forward with one of the solutions discussed here. But if we see that this can be made to work by patching the kernel we'll go that way.

**Comment by brauner** [comment]
So we're having a discussion to fix this in the kernel. There might even be a talk about this at LPC in Vancouver. We'll need to sort the details. The reason for blocking this is basically complexity.

**Comment by stgraber** [comment]
@brauner update on the approach and what's needed? Is it a case where we need the new VFS API?

**Comment by tcf909** [comment]
Did this ever get resolved? I'm running in to the same issue currently (2 years later).
